stdlib/path.gr

```diff
@@ -0,0 +1,775 @@
+ */
+enum Path {
+  AbsoluteFile(TypedPath<Absolute, File>),
+  AbsoluteDir(TypedPath<Absolute, Directory>),
+  RelativeFile(TypedPath<Relative, File>),
+  RelativeDir(TypedPath<Relative, Directory>),
+}
+
+/**
+ * Represents a platform-specific path encoding scheme.
+ */
+export enum Platform {
+  Windows,
+  Posix,
+}
+
+/**
+ * Represents an error that can occur when finding a property of a path.
+ */
+export enum PathOperationError {
+  IncompatiblePathType,
+}
+
+/**
+ * Represents an error that can occur when appending paths.
+ */
+export enum AppendError {
+  AppendToFile,
+  AppendAbsolute,
+}
+
+/**
+ * Represents the status of an ancestry check between two paths.
+ */
+export enum AncestryStatus {
+  Descendant,
+  Ancestor,
+  Self,
+  NoLineage,
+}
+
+/**
+ * Represents an error that can occur when the types of paths are incompatible
+ * for an operation.
+ */
+export enum IncompatibilityError {
+  DifferentRoots,
+  DifferentBases,
+}
+
+/**
+ * Represents possible errors for the `relativeTo` operation.
+ */
+export enum RelativizationError {
+  Incompatible(IncompatibilityError),
+  ImpossibleRelativization,
+}
+
+/**
+ * @section Values: Functions for working with Paths.
+ */
+
+let makeToken = str => {
+  match (str) {
+    "." => Dot,
+    ".." => Dotdot,
+    _ when String.length(str) == 2 && String.charAt(1, str) == ':' =>
+      DriveTok(String.charAt(0, str)),
+    _ => Text(str),
+  }
+}
+
+let parseNextToken = (path: PathInfo, nextToken) => {
+  let (base, ft, subs) = path
+  match (nextToken) {
+    Slash | Dot => path,
+    DriveTok(label) => (base, ft, [Char.toString(label) ++ ":", ...subs]),
+    Text(str) => (base, ft, [str, ...subs]),
+    Dotdot => {
+      match (path) {
+        (_, _, [_, ...rest]) => (base, ft, rest),
+        (Rel(upDirs), _, []) => (Rel(upDirs + 1), ft, []),
+        (Abs(_), _, []) => path,
+      }
+    },
+  }
+}
+
+// splits a path on forward slashes
+let lexPath = (pathStr, platform) => {
+  let isSeparator = char => {
+    char == '/' || platform == Windows && char == '\\'
+  }
+  let len = String.length(pathStr)
+  let mut revTokens = []
+  let mut segBeginI = 0
+  for (let mut i = 0; i < len; i += 1) {
+    if (isSeparator(String.charAt(i, pathStr))) {
+      if (segBeginI != i) {
+        let tok = makeToken(String.slice(segBeginI, i, pathStr))
+        revTokens = [tok, ...revTokens]
+      }
+      revTokens = [Slash, ...revTokens]
+      segBeginI = i + 1
+    }
+  }
+  if (segBeginI < len) {
+    let lastPart = String.slice(segBeginI, len, pathStr)
+    revTokens = [makeToken(lastPart), ...revTokens]
+  }
+  List.reverse(revTokens)
+}
+
+let isFilePath = tokens => {
+  let revTokens = List.reverse(tokens)
+  match (revTokens) {
+    [Dot | Dotdot | Slash, ..._] => false,
+    _ => true,
+  }
+}
+
+let parseAbs = (pathStr, fileType, platform) => {
+  match (lexPath(pathStr, platform)) {
+    [] => None,
+    [first, ...rest] => {
+      let tokens = [first, ...rest]
```

Comment:
```suggestion
    [first, ...rest] as tokens => {
```
I believe we can use `as` syntax here. I'm also wondering why it doesn't warn about not handling `[first]`... I assume `[first, ...rest]` matches any non-empty list and just makes `rest` an empty list if there is exactly 1 item.